src/odemis/acq/stitching/_registrar.py

```diff
@@ -746,4 +746,8 @@ def _assemble_mosaic(self):
             if i not in positions:
                 positions[i] = [None, None]
 
-        return numpy.array(list(positions.values())).reshape([num_rows, num_cols, 2])
+        # Sort positions by key
```

Comment:
I think you could entirely do without the dict, and that it would clarify what's happening. `positions` should be there for each int 0->len(tree) -1. So it's actually an array.
Also, you have this special code to handle missing positions, but I think that the caller doesn't handle `[None, None] ` afterwards anyway. So I'd suggest to start with the "final array", initialised at [0,0], and update it on-the-fly as you're following the graph.
```
positions = numpy.zeros((num_rows, num_cols, 2)) # start with no shift for each tile
positions_flat = positions.reshape((num_rows * num_cols, 2))  # the tiles in the "idx" notation
:
: the while loop as before, using positions_flat
: 
# No special handling of the missing tiles or dict sorting!
return positions
```

Reply:
I added a new commit for this, not sure though if it's making it much better. It seems that I used the dictionary, so I can easily test if a position has already been considered. Now, I have to keep an additional list for this.

Reply:
To make it even better, we could consider cleaning up adj_positions a bit earlier.

Reply:
But maybe it's good enough already, I don't really see the need to improve it any further.